In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

In [2]:
!pip install mlxtend

Defaulting to user installation because normal site-packages is not writeable
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.4/1.4 MB 5.3 MB/s eta 0:00:005.7 MB/s eta 0:00:01


In [3]:
import csv
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori, association_rules

In [4]:
data = []
with open('/home/yxsh/Desktop/dataset/Market_Basket_Optimisation.csv') as file:
    reader = csv.reader(file, delimiter=',')
    for row in reader:
        data +=[row]

In [5]:
data[1:10] #list of list

[['burgers', 'meatballs', 'eggs'],
 ['chutney'],
 ['turkey', 'avocado'],
 ['mineral water', 'milk', 'energy bar', 'whole wheat rice', 'green tea'],
 ['low fat yogurt'],
 ['whole wheat pasta', 'french fries'],
 ['soup', 'light cream', 'shallot'],
 ['frozen vegetables', 'spaghetti', 'green tea'],
 ['french fries']]

In [6]:
len(data)

7501

In [7]:
te = TransactionEncoder()
x = te.fit_transform(data)

In [8]:
x

array([[False,  True,  True, ...,  True, False, False],
       [False, False, False, ..., False, False, False],
       [False, False, False, ..., False, False, False],
       ...,
       [False, False, False, ..., False, False, False],
       [False, False, False, ..., False, False, False],
       [False, False, False, ..., False,  True, False]])

In [9]:
te.columns_

[' asparagus',
 'almonds',
 'antioxydant juice',
 'asparagus',
 'avocado',
 'babies food',
 'bacon',
 'barbecue sauce',
 'black tea',
 'blueberries',
 'body spray',
 'bramble',
 'brownies',
 'bug spray',
 'burger sauce',
 'burgers',
 'butter',
 'cake',
 'candy bars',
 'carrots',
 'cauliflower',
 'cereals',
 'champagne',
 'chicken',
 'chili',
 'chocolate',
 'chocolate bread',
 'chutney',
 'cider',
 'clothes accessories',
 'cookies',
 'cooking oil',
 'corn',
 'cottage cheese',
 'cream',
 'dessert wine',
 'eggplant',
 'eggs',
 'energy bar',
 'energy drink',
 'escalope',
 'extra dark chocolate',
 'flax seed',
 'french fries',
 'french wine',
 'fresh bread',
 'fresh tuna',
 'fromage blanc',
 'frozen smoothie',
 'frozen vegetables',
 'gluten free bar',
 'grated cheese',
 'green beans',
 'green grapes',
 'green tea',
 'ground beef',
 'gums',
 'ham',
 'hand protein bar',
 'herb & pepper',
 'honey',
 'hot dogs',
 'ketchup',
 'light cream',
 'light mayo',
 'low fat yogurt',
 'magazines',
 'mashe

In [10]:
df = pd.DataFrame(x, columns=te.columns_)

In [11]:
df

asparagus  almonds  antioxydant juice  asparagus  avocado  babies food  \
0          False     True               True      False     True        False   
1          False    False              False      False    False        False   
2          False    False              False      False    False        False   
3          False    False              False      False     True        False   
4          False    False              False      False    False        False   
...          ...      ...                ...        ...      ...          ...   
7496       False    False              False      False    False        False   
7497       False    False              False      False    False        False   
7498       False    False              False      False    False        False   
7499       False    False              False      False    False        False   
7500       False    False              False      False    False        False   

      bacon  barbecue sauce  black tea  blueberries  ...  turkey  \
0     False           False      False        False  ...   False   
1     False           False      False        False  ...   False   
2     False           False      False        False  ...   False   
3     False           False      False        False  ...    True   
4     False           False      False        False  ...   False   
...     ...             ...        ...          ...  ...     ...   
7496  False           False      False        False  ...   False   
7497  False           False      False        False  ...   False   
7498  False           False      False        False  ...   False   
7499  False           False      False        False  ...   False   
7500  False           False      False        False  ...   False   

      vegetables mix  water spray  white wine  whole weat flour  \
0               True        False       False              True   
1              False        False       False             False   
2              False        False       False             False   
3              False        False       False             False   
4              False        False       False             False   
...              ...          ...         ...               ...   
7496           False        False       False             False   
7497           False        False       False             False   
7498           False        False       False             False   
7499           False        False       False             False   
7500           False        False       False             False   

      whole wheat pasta  whole wheat rice   yams  yogurt cake  zucchini  
0                 False             False   True        False     False  
1                 False             False  False        False     False  
2                 False             False  False        False     False  
3                 False             False  False        False     False  
4                 False              True  False        False     False  
...                 ...               ...    ...          ...       ...  
7496              False             False  False        False     False  
7497              False             False  False        False     False  
7498              False             False  False        False     False  
7499              False             False  False        False     False  
7500              False             False  False         True     False  

[7501 rows x 120 columns]

In [12]:
# Finding frequent itemsets
freq_itemset = apriori(df, min_support=0.01, use_colnames=True)

0.01 means 1% of data size(7501) is 75. so single item should come atleast 75 times then take those items in freq_itemset

In [13]:
freq_itemset

support                                 itemsets
0    0.020397                                (almonds)
1    0.033329                                (avocado)
2    0.010799                         (barbecue sauce)
3    0.014265                              (black tea)
4    0.011465                             (body spray)
..        ...                                      ...
252  0.011065       (ground beef, milk, mineral water)
253  0.017064  (spaghetti, mineral water, ground beef)
254  0.015731         (spaghetti, milk, mineral water)
255  0.010265    (olive oil, spaghetti, mineral water)
256  0.011465     (pancakes, spaghetti, mineral water)

[257 rows x 2 columns]

In [14]:
#Find the rules
rules = association_rules(freq_itemset, metric='confidence', min_threshold=0.10)

In [15]:
rules = rules[['antecedents','consequents','support','confidence']]
rules

antecedents                 consequents   support  \
0                     (avocado)             (mineral water)  0.011598   
1                     (burgers)                      (cake)  0.011465   
2                        (cake)                   (burgers)  0.011465   
3                   (chocolate)                   (burgers)  0.017064   
4                     (burgers)                 (chocolate)  0.017064   
..                          ...                         ...       ...   
315                 (olive oil)  (spaghetti, mineral water)  0.010265   
316       (pancakes, spaghetti)             (mineral water)  0.011465   
317   (pancakes, mineral water)                 (spaghetti)  0.011465   
318  (spaghetti, mineral water)                  (pancakes)  0.011465   
319                  (pancakes)  (spaghetti, mineral water)  0.011465   

     confidence  
0      0.348000  
1      0.131498  
2      0.141447  
3      0.104150  
4      0.195719  
..          ...  
315    0.155870  
316    0.455026  
317    0.339921  
318    0.191964  
319    0.120617  

[320 rows x 4 columns]

In [16]:
rules[rules['antecedents'] == {'cake'}]['consequents']

2               (burgers)
25            (chocolate)
27                 (eggs)
29         (french fries)
31    (frozen vegetables)
32            (green tea)
35                 (milk)
36        (mineral water)
39             (pancakes)
41            (spaghetti)
Name: consequents, dtype: object